### PR TITLE
KAFKA-5738: Add cumulative count for rate metrics (KIP-187)

### PR DESCRIPTION
**Pull request by rajinisivaram**
Implementation of https://cwiki.apache.org/confluence/display/KAFKA/KIP-187+-+Add+cumulative+count+metric+for+all+Kafka+rate+metrics

Also made locking in Sensor for `CompoundStat` consistent with simple `Stat`, avoiding locking the whole sensor.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6831/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6845/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6847/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6833/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -829,9 +829,9 @@ public GroupCoordinatorMetrics(Metrics metrics, String metricGrpPrefix) {
             this.heartbeatLatency.add(metrics.metricName("heartbeat-response-time-max",
                 this.metricGrpName,
                 "The max time taken to receive a response to a heartbeat request"), new Max());
-            this.heartbeatLatency.add(metrics.metricName("heartbeat-rate",
-                this.metricGrpName,
-                "The average number of heartbeats per second"), new Rate(new Count()));
+            this.heartbeatLatency.add(new Meter(new Count(),
+                    metrics.metricName("heartbeat-rate", this.metricGrpName, "The average number of heartbeats per second"),
+                    metrics.metricName("heartbeat-total", this.metricGrpName, "The total number of heartbeats")));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I wonder if we could have a helper method where you passed the metric group name, the name prefix (i.e. `heartbeat`) and the description name ("heartbeats" in this case, but "group joins" in another example) to be used in the description.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetricsImpl.java [hunk 2]
```diff
@@ -171,8 +171,15 @@ private void addLatencyMetrics(String scopeName, Sensor sensor, String opName, M
...
-            "The average number of occurrence of " + opName + " operation per second.", tags), new Rate(new Count()));
+        MetricName rateMetricName = metrics.metricName(opName + "-rate", groupNameFromScope(scopeName),
+            "The average number of occurrence of " + opName + " operation per second.", tags);
+        MetricName totalMetricName = metrics.metricName(opName + "-total", groupNameFromScope(scopeName),
+                "The total number of occurrence of " + opName + " operations.", tags);
+        if (!metrics.metrics().containsKey(rateMetricName) && !metrics.metrics().containsKey(totalMetricName)) {
+            sensor.add(new Meter(new Count(), rateMetricName, totalMetricName));
+        } else {
+            log.trace("Trying to add metric twice: {} {}", rateMetricName, totalMetricName);
+        }
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetricsImpl.java:182]: cc @dguy @guozhangwang

File: clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java [hunk 3]
```diff
@@ -224,8 +224,9 @@ public void add(CompoundStat stat) {
      */
     public synchronized void add(CompoundStat stat, MetricConfig config) {
         this.stats.add(Utils.notNull(stat));
+        Object lock = new Object();
         for (NamedMeasurable m : stat.stats()) {
-            KafkaMetric metric = new KafkaMetric(this, m.name(), m.stat(), config == null ? this.config : config, time);
+            KafkaMetric metric = new KafkaMetric(lock, m.name(), m.stat(), config == null ? this.config : config, time);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:229]: Good to mention this change in the PR description since it's an unrelated bug fix.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 1 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Done (for the consumer).

File: clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java (hunk 3 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:229]: Done.

**Comment by rajinisivaram** [comment]
@ijuma Thanks for the review. I have updated, rebased and added unit tests.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java [hunk 4]
```diff
@@ -858,46 +858,56 @@ public SelectorMetrics(Metrics metrics, String metricGrpPrefix, Map<String, Stri
             }
 
             this.connectionClosed = sensor("connections-closed:" + tagsSuffix.toString());
-            MetricName metricName = metrics.metricName("connection-close-rate", metricGrpName, "Connections closed per second in the window.", metricTags);
-            this.connectionClosed.add(metricName, new Rate());
+            MetricName rateMetricName = metrics.metricName("connection-close-rate", metricGrpName, "Connections closed per second in the window.", metricTags);
+            MetricName totalMetricName = metrics.metricName("connection-close-total", metricGrpName, "Total count of connections closed.", metricTags);
+            this.connectionClosed.add(new Meter(rateMetricName, totalMetricName));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Can we not use the same approach we used in the consumer code?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -492,31 +493,45 @@ StandbyTask createTask(final Consumer<byte[], byte[]> consumer, final TaskId tas
             commitTimeSensor = metrics.sensor(prefix + ".commit-latency", Sensor.RecordingLevel.INFO);
             commitTimeSensor.add(metrics.metricName("commit-latency-avg", this.groupName, "The average commit time in ms", this.tags), new Avg());
             commitTimeSensor.add(metrics.metricName("commit-latency-max", this.groupName, "The maximum commit time in ms", this.tags), new Max());
-            commitTimeSensor.add(metrics.metricName("commit-rate", this.groupName, "The average per-second number of commit calls", this.tags), new Rate(new Count()));
+            MetricName rateMetricName = metrics.metricName("commit-rate", this.groupName, "The average per-second number of commit calls", this.tags);
+            MetricName totalMetricName = metrics.metricName("commit-total", this.groupName, "The total number of commit calls", this.tags);
+            commitTimeSensor.add(new Meter(new Count(), rateMetricName, totalMetricName));
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Here too, can we not use similar approach as we used in the consumer?

**Review by ijuma — COMMENTED** [review]
Thanks for the update, did another pass, just a couple of minor comments.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java (hunk 4 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: I was a bit concerned about this because the descriptions were not very consistent across metrics and I didn't want to lose anything from the description. Have updated.

**Comment by rajinisivaram** [comment]
@ijuma Thank you for the review. Updated the code.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 6]
```diff
@@ -492,34 +494,42 @@ StandbyTask createTask(final Consumer<byte[], byte[]> consumer, final TaskId tas
             commitTimeSensor = metrics.sensor(prefix + ".commit-latency", Sensor.RecordingLevel.INFO);
             commitTimeSensor.add(metrics.metricName("commit-latency-avg", this.groupName, "The average commit time in ms", this.tags), new Avg());
             commitTimeSensor.add(metrics.metricName("commit-latency-max", this.groupName, "The maximum commit time in ms", this.tags), new Max());
-            commitTimeSensor.add(metrics.metricName("commit-rate", this.groupName, "The average per-second number of commit calls", this.tags), new Rate(new Count()));
+            commitTimeSensor.add(createMeter(metrics, new Count(), "commit", "commit calls"));
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:497]: Nice when we add functionality _and_ make the code more concise. :)

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 7]
```diff
@@ -775,28 +779,32 @@ private void maybeRegisterTopicMetrics(String topic) {
...
 
                 String topicByteRateName = "topic." + topic + ".bytes";
                 Sensor topicByteRate = this.metrics.sensor(topicByteRateName);
-                m = this.metrics.metricInstance(metricsRegistry.topicByteRate, metricTags);
-                topicByteRate.add(m, new Rate());
+                rateMetricName = this.metrics.metricInstance(metricsRegistry.topicByteRate, metricTags);
+                totalMetricName = this.metrics.metricInstance(metricsRegistry.topicByteTotal, metricTags);
+                topicByteRate.add(new Meter(rateMetricName, totalMetricName));
 
                 String topicCompressionRateName = "topic." + topic + ".compression-rate";
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:788]: This should really be compression-ratio. Annoying.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    public Meter(TimeUnit unit, MetricName rateMetricName, MetricName totalMetricName) {
+        this(unit, new SampledTotal(), rateMetricName, totalMetricName);
+    }
+
+    public Meter(SampledStat total, MetricName rateMetricName, MetricName totalMetricName) {
+        this(TimeUnit.SECONDS, total, rateMetricName, totalMetricName);
+    }
+
+    public Meter(TimeUnit unit, SampledStat sampledStat, MetricName rateMetricName, MetricName totalMetricName) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: Nit: should we call this `rateUnit`? Same for the other constructor.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    public Meter(MetricName rateMetricName, MetricName totalMetricName) {
+        this(TimeUnit.SECONDS, new SampledTotal(), rateMetricName, totalMetricName);
+    }
+
+    public Meter(TimeUnit unit, MetricName rateMetricName, MetricName totalMetricName) {
+        this(unit, new SampledTotal(), rateMetricName, totalMetricName);
+    }
+
+    public Meter(SampledStat total, MetricName rateMetricName, MetricName totalMetricName) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: The other constructor calls the parameter `sampledStat`. We should be consistent.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        this.rateMetricName = rateMetricName;
+        this.totalMetricName = totalMetricName;
+    }
+
+    @Override
+    public List<NamedMeasurable> stats() {
+        List<NamedMeasurable> stats = new ArrayList<NamedMeasurable>(2);
+        stats.add(new NamedMeasurable(totalMetricName, total));
+        stats.add(new NamedMeasurable(rateMetricName, rate));
+        return stats;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: Can we use `Arrays.asList`?

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    public Meter(TimeUnit unit, SampledStat sampledStat, MetricName rateMetricName, MetricName totalMetricName) {
+        this.total = new Total();
+        this.rate = new Rate(unit, sampledStat) {
+            @Override
+            public void record(MetricConfig config, double value, long now) {
+                super.record(config, value, now);
+                total.record(config, value, now);
+            }
+        };
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: Why don't we simply do these two calls from `record` without overriding `Rate`?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 12]
```diff
@@ -1508,6 +1508,54 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     servers.foreach(assertNoExemptRequestMetric(_))
   }
 
+  @Test
+  def testRateMetricsHaveCumulativeCount() {
+    val numRecords = 1000
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Would 100 records be enough for this test?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 13]
```diff
@@ -1508,6 +1508,54 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+    sendRecords(numRecords)
+
+    val consumer = this.consumers.head
+    consumer.assign(List(tp).asJava)
+    consumer.seek(tp, 0)
+    consumeAndVerifyRecords(consumer, numRecords = numRecords, startingOffset = 0)
+
+    def exists(name: String, metricName: MetricName, allMetricNames: Set[MetricName]): Boolean = {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Maybe `metricName` should be `rateMetricName` to make it clear why we create a new `MetricName` below.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 14]
```diff
@@ -1508,6 +1508,54 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     servers.foreach(assertNoExemptRequestMetric(_))
   }
 
+  @Test
+  def testRateMetricsHaveCumulativeCount() {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Maybe worth adding a method comment saying that we test both the producer and consumer here.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 15]
```diff
@@ -1508,6 +1508,54 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+
+    val producer = this.producers.head
+    val producerMetricNames = producer.metrics.keySet.asScala.toSet
+    val producerExclusions = Set("compression-rate") // compression-rate is an Average metric, not Rate
+    producerMetricNames.filter(_.name.endsWith("-rate"))
+        .filterNot(metricName => producerExclusions.contains(metricName.name))
+        .foreach(verify(_, producerMetricNames))
+
+    // Check a couple of metrics to ensure that values are set
+    def metric(name: String, metrics: Map[MetricName, Metric]): (MetricName, Metric) =
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Maybe this can take `java.util.Map` so that the callers are more concise.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 16]
```diff
@@ -1508,6 +1508,54 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+      metrics.filter(e => e._1.name.equals(name)).head
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: `metrics.find { case (metricName, _) => metricName.name == name }`

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+ * A compound stat that includes a rate metric and a cumulative total metric.
+ */
+public class Meter implements CompoundStat {
+
+    private final MetricName rateMetricName;
+    private final MetricName totalMetricName;
+    private final Rate rate;
+    private final Total total;
+
+    public Meter(MetricName rateMetricName, MetricName totalMetricName) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: We should document the constructors mentioning the default choices they make for the user.

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/MeterTest.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+        assertEquals(2, stats.size());
+        NamedMeasurable rate;
+        NamedMeasurable total;
+        if (stats.get(0).name().equals(rateMetricName)) {
+            rate = stats.get(0);
+            total = stats.get(1);
+        } else {
+            rate = stats.get(1);
+            total = stats.get(0);
+        }
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/MeterTest.java]: Do we need the if/else? Since this is a unit test, it seems OK to just assert that the first element is the rate and the second is the total.

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/MeterTest.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+        Total totalStat = (Total) total.stat();
+
+        MetricConfig config = new MetricConfig();
+        double nextValue = 0.0;
+        double expectedTotal = 0.0;
+        long now = 0;
+        double intervalMs = 100;
+        double delta = 5.0;
+
+        for (int i = 1; i <= 100; i++) {
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/MeterTest.java]: It may be worth adding a comment explaining what we're trying to achieve here.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java [hunk 20]
```diff
@@ -910,6 +909,32 @@ public double measure(MetricConfig config, long now) {
...
+            else
+                return new Meter(stat, rateMetricName, totalMetricName);
+        }
+
+        private Meter createMeter(Metrics metrics, String groupName,  Map<String, String> metricTags,
+                String baseName, String descriptiveName) {
+            return createMeter(metrics, groupName, metricTags, null, baseName, descriptiveName);
+        }
+
+        private Meter createRatioMeter(Metrics metrics, String groupName,  Map<String, String> metricTags,
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Should probably be `createIoThreadRatio` since we mention `I/O thread` in the messages.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java [hunk 21]
```diff
@@ -860,46 +861,44 @@ public SelectorMetrics(Metrics metrics, String metricGrpPrefix, Map<String, Stri
...
-            metricName = metrics.metricName("outgoing-byte-rate", metricGrpName, "The average number of outgoing bytes sent per second to all servers.", metricTags);
-            this.bytesSent.add(metricName, new Rate());
-            metricName = metrics.metricName("request-rate", metricGrpName, "The average number of requests sent per second.", metricTags);
-            this.bytesSent.add(metricName, new Rate(new Count()));
-            metricName = metrics.metricName("request-size-avg", metricGrpName, "The average size of all requests in the window..", metricTags);
+            this.bytesSent.add(createMeter(metrics, metricGrpName, metricTags,
+                    "outgoing-byte", "outgoing bytes sent to all servers"));
+            this.bytesSent.add(createMeter(metrics, metricGrpName, metricTags, new Count(),
+                    "request", "requests sent"));
+            MetricName metricName = metrics.metricName("request-size-avg", metricGrpName, "The average size of all requests in the window..", metricTags);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Can we remove the extra full stop in the description? Also, does it still make sense to say "in the window" for this and a few others given that we don't seem to say it for most of them?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Looks good overall, just a few minor comments. Also, can we mention the KIP in the PR title? I usually add it like `(KIP-XX)` at the end.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java (hunk 9 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: Renamed to `rateStat`

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java (hunk 8 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java]: Renamed both to `rateStat`.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 12 above)
Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: yes, updated.

**Comment by rajinisivaram** [comment]
@ijuma Thank you for the review. Have addressed the comments.

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/MeterTest.java [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+            for (; now < i * 1000; now += intervalMs, nextValue += delta) {
+                expectedTotal += nextValue;
+                meter.record(config, nextValue, now);
+            }
+            assertEquals(expectedTotal, totalStat.measure(config, now), EPS);
+            long windowSizeMs = rateStat.windowSize(config, now);
+            long windowStartMs = Math.max(now - windowSizeMs, 0);
+            double sampledTotal = 0.0;
+            double prevValue = nextValue - delta;
+            for (long timeMs = now - 100; timeMs >= windowStartMs; timeMs -= intervalMs, prevValue -= delta)
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/MeterTest.java:68]: Does `100` represent `intervalMs` here?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetricsImplTest.java [hunk 23]
```diff
@@ -100,8 +102,10 @@ public void testThroughputMetrics() {
         Sensor sensor1 = streamsMetrics.addThroughputSensor(scope, entity, operation, Sensor.RecordingLevel.DEBUG);
 
         Map<MetricName, ? extends Metric> metrics = streamsMetrics.metrics();
-        // 2 metrics plus a common metric that keeps track of total registered metrics in Metrics() constructor
-        assertEquals(metrics.size(), 3);
+        int meterMetricsCount = 2; // Each Meter is a combination of a Rate and a Total
+        int otherMetricsCount = 0;
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetricsImplTest.java:106]: I think we don't need this since there are 0.

**Review by ijuma — APPROVED** [review]
Thanks for the updates, LGTM. Merging to trunk.
